### PR TITLE
Add some log info within GradientTape

### PR DESCRIPTION
**Pull request by a6802739**
@alextp, reference to #26143.

I saw @shashvatshahi1998 made some changes already, So I did some supplement here.

Please take a look at it, Thanks.

### CODE REVIEW

File: tensorflow/python/eager/backprop.py [hunk 1]
```diff
@@ -811,6 +827,9 @@ def watch(self, tensor):
     Args:
       tensor: a Tensor or list of Tensors.
     """
+    if check_float_dtype(tensor) == False:
```
Comment (superbobry) [inline tensorflow/python/eager/backprop.py]: You can fuse `check_float_dtype` into the for loop. This will allow to issue a more specific warning (see comment below):

```python
for t in tf.nest.flatten(tensor):
  if not t.dtype.is_floating:
    ...
```

File: tensorflow/python/eager/backprop.py [hunk 2]
```diff
@@ -811,6 +827,9 @@ def watch(self, tensor):
...
+        logging.log_first_n(logging.WARN, "The expected Datatype must be 'tf.float' "
```
Comment (superbobry) [inline tensorflow/python/eager/backprop.py]: Is there a specific reason to log just on the first occurrence? I think logging for every `gradient` call is OK.

Consider making the error message a bit more specific, something like

```python
"The expected dtype of the watched tensor must be floating (e.g. tf.float32), got {!r}".format(t.dtype)
```

File: tensorflow/python/eager/backprop.py [hunk 3]
```diff
@@ -938,6 +957,9 @@ def gradient(self,
             "gradient in order to compute higher order "
             "derivatives.", 1)
 
+    if check_float_dtype(target) == False:
```
Comment (superbobry) [inline tensorflow/python/eager/backprop.py]: Same comment as above: fuse into the loop and make the error message more specific.

**Review by superbobry — CHANGES_REQUESTED** [review]
Thanks for doing that!

File: tensorflow/python/eager/backprop.py (hunk 3 above)
Comment (superbobry) [reply tensorflow/python/eager/backprop.py]: Also add a similar check to `sources` below.

File: tensorflow/python/eager/backprop.py (hunk 1 above)
Comment (a6802739) [reply tensorflow/python/eager/backprop.py]: Thanks for your review. Done now.

File: tensorflow/python/eager/backprop.py (hunk 2 above)
Comment (a6802739) [reply tensorflow/python/eager/backprop.py]: Yeah, just consider maybe too much warning there, just change it as you say. Thanks.

File: tensorflow/python/eager/backprop.py (hunk 3 above)
Comment (a6802739) [reply tensorflow/python/eager/backprop.py]: Done.

**Comment by shashvatshahi1998** [comment]
@superbobry please review my pull request also I don't know why my cla is still showing no while Ihave signed that on 8th March

File: tensorflow/python/eager/backprop.py [hunk 4]
```diff
@@ -812,6 +812,9 @@ def watch(self, tensor):
       tensor: a Tensor or list of Tensors.
     """
     for t in nest.flatten(tensor):
+      if not t.dtype.is_floating:
+        logging.vlog(logging.WARN, "The expected dtype of the watched tensor must be "
+        "floating (e.g. tf.float32), got {!r}".format(t.dtype))
```
Comment (superbobry) [inline tensorflow/python/eager/backprop.py]: 4 spaces for continuation indent, please, e.g.

```python
logging.vlog("..."
    "...")
```

Also (minor nit): I think it's fine to drop "expected" from the error messages.

Comment (a6802739) [reply tensorflow/python/eager/backprop.py]: Done, Thanks!

**Review by superbobry — DISMISSED** [review]

**Comment by alextp** [comment]
Please fix the lint errors:

```
tensorflow/python/eager/backprop.py:950: [C0301(line-too-long), ] Line too long (85/80)

tensorflow/python/eager/backprop.py:964: [C0301(line-too-long), ] Line too long (85/80)
```

**Comment by a6802739** [comment]
@alextp, Done now. Thanks!

File: tensorflow/python/eager/backprop.py [hunk 5]
```diff
@@ -812,6 +812,10 @@ def watch(self, tensor):
       tensor: a Tensor or list of Tensors.
     """
     for t in nest.flatten(tensor):
+      if not t.dtype.is_floating:
+        logging.vlog(
+            logging.WARN, "The dtype of the watched tensor must be "
+            "floating (e.g. tf.float32), got {!r}".format(t.dtype))
```
Comment (superbobry) [inline tensorflow/python/eager/backprop.py]: Sorry, one more nitpick: could you replace vlog(..., "... {!r}".format(t.dtype)) with vlog(..., "... %r", t.dtype) so that interpolation is done lazily? Same for other messages.

Comment (a6802739) [reply tensorflow/python/eager/backprop.py]: Done now, Thanks!

**Comment by a6802739** [comment]
@alextp  @superbobry, Thanks for all you help!

**Review by alextp — APPROVED** [review]

**Comment by a6802739** [comment]
@alextp, Appreciate you help again, and why the check is pending now?

**Comment by a6802739** [comment]
@alextp, why `Ubuntu Makefile` failed? I couldn't see any details about it.

**Comment by alextp** [comment]
No idea, I can't see it either. Rerunning.

**Comment by a6802739** [comment]
@alextp, It it have already be merged, I think this thread could be closed now.
